internal/gke/creds.go

```diff
@@ -0,0 +1,423 @@
+	if err != nil {
+		return "", err
+	}
+	var creds Creds
+
+	if err := json.Unmarshal([]byte(ecStr), &creds); err != nil {
+		return "", fmt.Errorf("unable to retrieve credentials: %w", err)
+	}
+
+	return creds.Status.Token, nil
+}
+
+// execCredential return an object of type ExecCredential which
+// holds a bearer token to authenticate to GKE.
+func (p *plugin) execCredential() (*clientauthv1b1.ExecCredential, error) {
+	token, expiry, err := p.accessToken()
+	if err != nil {
+		return nil, err
+	}
+
+	return &clientauthv1b1.ExecCredential{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ExecCredential",
+			APIVersion: "client.authentication.k8s.io/v1beta1",
+		},
+		Status: &clientauthv1b1.ExecCredentialStatus{
+			Token:               token,
+			ExpirationTimestamp: expiry,
+		},
+	}, nil
+}
+
+// accessToken return either the ApplicationDefaultCredentials or the gcloudAccessToken
+func (p *plugin) accessToken() (string, *metav1.Time, error) {
+	if p.useApplicationDefaultCredentials {
+		return p.defaultAccessToken()
+	}
+	return p.gcloudAccessToken()
+}
+
+// gcloudAccessToken returns a cached token if the token is not expired. If the token is
+// expired, it gets a new access token by invoking gcloud command, caches the new token
+// and returns the token.
+func (p *plugin) gcloudAccessToken() (string, *metav1.Time, error) {
+	if token, expiry, err := p.getCachedGcloudAccessToken(); err != nil {
+		// log and ignore error; move on to getting a new token from gcloud
+		klog.V(4).Infof("Getting cached gcloud access token failed with error: %v", err)
+	} else {
+		// return valid token
+		return token, expiry, nil
+	}
+
+	gc, err := p.readGcloudConfig()
+	if err != nil {
+		return "", nil, err
+	}
+
+	if gc.Credential.AccessToken == "" {
+		return "", nil, fmt.Errorf("gcloud config config-helper returned an empty access token")
+	}
+	if gc.Credential.TokenExpiry.IsZero() {
+		return "", nil, fmt.Errorf("failed to retrieve expiry time from gcloud config json object")
+	}
+
+	// Authorization Token File is not commonly used. Currently, this is for specific internal debugging scenarios.
+	token := gc.Credential.AccessToken
+	var authzTokenFile string
+	var authzTokenBytes []byte
+	if authzTokenFile = gc.Configuration.Properties.Auth.AuthorizationTokenFile; authzTokenFile != "" {
+		authzTokenBytes, err = p.readFile(authzTokenFile)
+		if err != nil {
+			return "", nil, fmt.Errorf("gcloud config sets property auth/authorization_token_file, but can't read file at %s: %w", authzTokenFile, err)
+		}
+		token = fmt.Sprintf("iam-%s^%s", token, authzTokenBytes)
+	}
+
+	if err := p.writeGcloudAccessTokenToCache(token, gc.Credential.TokenExpiry); err != nil {
+		// log and ignore error as writing to cache is best effort
+		klog.V(4).Infof("Failed to write gcloud access token to cache with error: %v", err)
+	}
+
+	return token, &metav1.Time{Time: gc.Credential.TokenExpiry}, nil
+}
+
+func (p *plugin) defaultAccessToken() (string, *metav1.Time, error) {
```

Comment:
I think that one is actually the only thing we need to do to get the actual token for kubeconfig

Reply:
Actually, I've found the below that might be also some alternative:
https://pkg.go.dev/google.golang.org/api/idtoken